### PR TITLE
[url_launcher] Fix crashes on a lot of devices

### PR DESCRIPTION
**Pull request by fzyzcjy**
Please see https://github.com/flutter/flutter/issues/72863 for **a lot of** people facing this issue making the app crash.

*List which issues are fixed by this PR. You must list at least one issue.*
Closes https://github.com/flutter/flutter/issues/72863

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*
N/A

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I [updated pubspec.yaml](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates) with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code

### CODE REVIEW

**Comment by stuartmorgan** [comment]
Thanks for the submission! I'm closing in favor of the existing https://github.com/flutter/plugins/pull/4174, which is further along; feel free to suggest changes there.

In the future, please don't check boxes in the checklist for steps you haven't completed. The checklist is there for a reason. Ignoring it will only slow reviews and waste developer time.

**Comment by fzyzcjy** [comment]
@stuartmorgan Sure. Sorry for that.

**Comment by fzyzcjy** [comment]
@stuartmorgan Hi, I see that PR not merged after months. On the other hand, my PR only contains 3 lines of code (indeed only 1 line useful). So maybe this can be reopened?

**Comment by stuartmorgan** [comment]
If you're interested in completing the PR I'm happy to re-open it. Since it's incomplete I'll mark it as a draft; please remove that designation once you've actually worked through the PR checklist.

**Comment by fzyzcjy** [comment]
@stuartmorgan IMHO I am done now.

**Comment by fzyzcjy** [comment]
oops wait a minute

**Comment by stuartmorgan** [comment]
> @stuartmorgan IMHO I am done now.

Where is the new test you are indicating here:

> * [x]  I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.

that you have added? It does not appear to have been pushed to your PR branch.

File: packages/url_launcher/url_launcher/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+## 6.0.13
+
+* Fix crashes on a lot of devices.
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher/CHANGELOG.md:3]: This is extremely vague; please describe the issue being fixed. You haven't mentioned either the platform (even though this is platform-specific), nor the circumstances.

**Comment by fzyzcjy** [comment]
@stuartmorgan IMHO this is quite safe so test exempt; but if needed I can add a test.

File: packages/url_launcher/url_launcher/CHANGELOG.md (hunk 1 above)
Comment (fzyzcjy) [reply packages/url_launcher/url_launcher/CHANGELOG.md]: @stuartmorgan Done. Looking at https://github.com/flutter/flutter/issues/72863 , seems that nobody has a clear circumstance - it is just "many people see their customers have this crash but do not know how and why".

Comment (stuartmorgan) [reply packages/url_launcher/url_launcher/CHANGELOG.md]: https://github.com/flutter/flutter/issues/72863#issuecomment-777983748 seems pretty specific.

**Comment by stuartmorgan** [comment]
> @stuartmorgan IMHO this is quite safe so test exempt; but if needed I can add a test.

You've indicated in the checklist that you've read the contributing guide, which endeavors to be very clear on this topic. Could you point me to the specific section that made you believe this PR would be exempt so that we can improve it?

File: packages/url_launcher/url_launcher/CHANGELOG.md (hunk 1 above)
Comment (fzyzcjy) [reply packages/url_launcher/url_launcher/CHANGELOG.md]: @stuartmorgan done

**Comment by fzyzcjy** [comment]
@stuartmorgan Could you please provide some suggestions on how I should test it?

Firstly, https://github.com/flutter/plugins/blob/master/packages/url_launcher/url_launcher/android/src/test/java/io/flutter/plugins/urllauncher/MethodCallHandlerImplTest.java seems only to deal with MethodCallHandlerImpl, so should not put here.

Secondly, there is no such things as WebViewActivityTest.java. And imho, if those 189 lines of code have no tests, should my 3 (indeed 1) lines of code have tests? Espeically considering that, this is just **defensive programming**, not useful real logic.

Thirdly, it seems also not appropreate to write test in dart, as this is quite android specific.

**Comment by stuartmorgan** [comment]
> @stuartmorgan Could you please provide some suggestions on how I should test it?

I would use a Java unit test, probably extracting a utility method to configure a webview from an intent and then unit testing that.

> Secondly, there is no such things as WebViewActivityTest.java.

Making a new file is trivial.

> And imho, if those 189 lines of code have no tests, should my 3 (indeed 1) lines of code have tests?

https://github.com/flutter/flutter/wiki/Plugin-Tests#faq-do-i-need-to-add-tests-even-if-the-part-of-the-code-im-changing-isnt-already-tested

You've repeatedly cited your own opinion as justification for not testing; the criteria for test exemptions are clearly laid out in the documents you've indicated that you've read, and that's not one of them.

Again, the documentation and the checklist exist to avoid rehashing things in every PR, as that is not a good use of anyone's time. If you want to argue for the Flutter testing project to change its testing policy, this isn't the place for it..

> Espeically considering that, this is just **defensive programming**, not useful real logic.

Not crashing seems useful to me; the fact that you are submitting a PR suggests that you believe it's useful as well. Therefore it needs tests to avoid regressing in the future.

> Thirdly, it seems also not appropreate to write test in dart, as this is quite android specific.

There's nothing wrong with doing platform-specific tests in Dart, but in this particular case I don't see a way to do that.

**Comment by stuartmorgan** [comment]
Closing in favor of https://github.com/flutter/plugins/pull/4507, which has working tests, and has passed review.

**Pull request closed without merging**